SshTool choice

aledsage opened the pull request:
Adds SshMachineLocation.SSH_TOOL_CLASS
    
- Deprecates SshTool.PROP_TOOL_CLASS
- Adds tests

bostko commented on core/src/main/java/org/apache/brooklyn/util/core/internal/ssh/SshTool.java:
```diff
@@ -49,7 +50,14 @@
+     * @deprecated since 0.9.0; use {@link SshMachineLocation#SSH_TOOL_CLASS}
+     */
+    @Deprecated
+    public static final ConfigKey<String> PROP_TOOL_CLASS = newStringConfigKey(
```
Can you explain why did you deprecate this?
It is good if this key is in a generic location like `SshTool`. I was thinking about refactoring Winrm support to follow the same pattern as it is for ssh, if so we could implement this dynamic sshTool loading with the same keys and the same API.
In regard with this please check https://github.com/apache/incubator-brooklyn/pull/950

ahgittin replied:
agree wtih @bostko about pattern where we share the static config key definition; but this particular key name is horrible so +1 to the deprecation.

aledsage replied:
We need a different key name for the class name of the WinRmTool and the SshTool. We'll use this by writing some configuration on the location (e.g. in brooklyn.properties on a named location), and will want to supply an implementation of WinRmTool and an implementation of SshTool - hence we need a different key name for each.

My reasoning for not having it in SshTool is that the config there is about config the SshTool understands. By the time you have an SshTool, then it has been instantiated. The place that instantiates the tool is SshMachineLocation, so that feels like a good place to define this config key. I also defined it in BrooklynConfigKeys, but with the common prefix "brooklyn.ssh.config.".

neykov commented on core/src/main/java/org/apache/brooklyn/location/localhost/LocalhostMachineProvisioningLocation.java:
```diff
@@ -73,6 +73,7 @@
 
     private static final long serialVersionUID = -7791239672433897762L;
 
+    /** @deprecated since 0.9.0; shouldn't be public */
```
Add `@Deprecated` annotation as well?

neykov commented on core/src/main/java/org/apache/brooklyn/location/ssh/SshMachineLocation.java:
```diff
@@ -135,13 +136,17 @@
 public class SshMachineLocation extends AbstractLocation implements MachineLocation, PortSupplier, WithMutexes, Closeable {
 
     /** @deprecated since 0.7.0 shouldn't be public */
```
Remove `@deprecated` marker now that these are private.

neykov commented:
LGTM, minor comments.

iyovcheva commented:
Looks good to me as well.

ahgittin commented on core/src/main/java/org/apache/brooklyn/core/config/ConfigKeys.java:
```diff
@@ -157,6 +157,14 @@ public static PortAttributeSensorAndConfigKey newPortSensorAndConfigKey(String n
         return newConfigKeyRenamed(prefix+key.getName(), key);
     }
 
+    public static <T> ConfigKey<T> newConfigKeyWithoutPrefix(String prefix, ConfigKey<T> key) {
```
`newConfigKeyWithPrefixRemoved` ?

ahgittin commented:
agree with others, minor comments, then good to merge

aledsage commented:
Thanks all. I've incorporated this feedback (and added comments to the code about different keys for WinRmTool versus SshTool implementations, and about why the config key is not to be defined in SshTool).

I've also rebased against master.

Waiting for jenkins to confirm that it builds, and then will merge.
